### ISSUE TITLE
LLVM15 does not like our calling convention scheme

### ISSUE THREAD

**Issue by Keno** [JuliaLang/julia#45057]
On LLVM15, during bootstrap, LLVM turns the following:
```
@MUL.CoreDOT.nothing20 = external dso_local global ptr
@MUL.CoreDOT.typeassert104 = external dso_local global ptr
@MUL.CoreDOT.tuple268 = external dso_local global ptr
@SUM.CoreDOT.String59 = external dso_local global ptr
@MUL.CoreDOT.String58 = external dso_local global ptr
@MUL.MainDOT.BaseDOT.print_to_string2339 = external dso_local global ptr
@jl_globalYY.2469 = external dso_local global ptr
@MUL.MainDOT.BaseDOT.SysDOT.BINDIR5428 = external dso_local global ptr
@MUL.MainDOT.BaseDOT.FilesystemDOT.joinpath3906 = external dso_local global ptr
@MUL.MainDOT.BaseDOT.FilesystemDOT.abspath4476 = external dso_local global ptr
@MUL.MainDOT.BaseDOT.SysDOT.STDLIB5135 = external dso_local global ptr
@jl_globalYY.5429 = external dso_local global ptr
@jl_globalYY.4888 = external dso_local global ptr
@jl_globalYY.5430 = external dso_local global ptr
@jl_globalYY.4890 = external dso_local global ptr
@jl_globalYY.5431 = external dso_local global ptr
@jl_globalYY.5432 = external dso_local global ptr
@jl_globalYY.5433 = external dso_local global ptr
@jl_globalYY.5434 = external dso_local global ptr
@_j_str7 = external dso_local constant [11 x i8]
@jlplt_ijl_get_julia_bindir_28472_got = external dso_local global ptr

declare ptr @julia.get_pgcstack()

; Function Attrs: cold noreturn nounwind
declare void @llvm.trap() #0

declare void @ijl_type_error()

declare ptr addrspace(10) @julia.typeof(ptr addrspace(10))

declare void @julia.write_barrier_binding(ptr addrspace(10), ...)

; Function Attrs: nocallback nofree nosync nounwind readnone speculatable willreturn
declare i32 @llvm.bswap.i32(i32) #1

define internal ptr addrspace(10) @japi1_print_to_string_32888() {
after_throw24:
  ret ptr addrspace(10) null
}

declare ptr addrspace(10) @julia_joinpath_24652()

define internal void @julia___init_build_28469(ptr %jlplt_ijl_get_julia_bindir_28472_got, ptr %SUM.CoreDOT.String59, ptr %0, ptr addrspace(10) %1, i1 %2, ptr %MUL.MainDOT.BaseDOT.SysDOT.BINDIR5428, ptr %3, ptr addrspace(10) %4, ptr %jl_globalYY.2469, ptr %5, ptr %jl_globalYY.5429, ptr %6, ptr %jl_globalYY.4888, ptr %7, ptr %jl_globalYY.5430, ptr %8, ptr %jl_globalYY.4890, ptr %9) {
top:
  %10 = call ptr @julia.get_pgcstack()
  %11 = load atomic ptr, ptr %jlplt_ijl_get_julia_bindir_28472_got unordered, align 8
  %12 = call ptr addrspace(10) %11()
  %13 = load ptr, ptr %jlplt_ijl_get_julia_bindir_28472_got, align 8
  %14 = addrspacecast ptr %jlplt_ijl_get_julia_bindir_28472_got to ptr addrspace(10)
  %15 = call ptr addrspace(10) @julia.typeof(ptr addrspace(10) null)
  %16 = icmp eq ptr addrspace(10) null, %1
  br i1 %2, label %pass, label %fail

fail:                                             ; preds = %top
  call void @ijl_type_error()
  unreachable

pass:                                             ; preds = %top
  %17 = load ptr, ptr %jlplt_ijl_get_julia_bindir_28472_got, align 8
  %18 = addrspacecast ptr %jlplt_ijl_get_julia_bindir_28472_got to ptr addrspace(10)
  call void (ptr addrspace(10), ...) @julia.write_barrier_binding(ptr addrspace(10) null, ptr addrspace(10) %1)
  %19 = load ptr, ptr %jlplt_ijl_get_julia_bindir_28472_got, align 8
  %20 = addrspacecast ptr %jlplt_ijl_get_julia_bindir_28472_got to ptr addrspace(10)
  %21 = load ptr, ptr %jlplt_ijl_get_julia_bindir_28472_got, align 8
  %22 = addrspacecast ptr %jlplt_ijl_get_julia_bindir_28472_got to ptr addrspace(10)
  %23 = load ptr, ptr %jlplt_ijl_get_julia_bindir_28472_got, align 8
  %24 = addrspacecast ptr %jlplt_ijl_get_julia_bindir_28472_got to ptr addrspace(10)
  %25 = load ptr, ptr %jlplt_ijl_get_julia_bindir_28472_got, align 8
  %26 = addrspacecast ptr %jlplt_ijl_get_julia_bindir_28472_got to ptr addrspace(10)
  %27 = load ptr, ptr %jlplt_ijl_get_julia_bindir_28472_got, align 8
  %28 = addrspacecast ptr %jlplt_ijl_get_julia_bindir_28472_got to ptr addrspace(10)
  %29 = call cc37 ptr addrspace(10) @japi1_print_to_string_32888(ptr addrspace(10) null, ptr addrspace(10) null, ptr addrspace(10) null, ptr addrspace(10) null, ptr addrspace(10) null)
  ret void
}

declare ptr addrspace(10) @julia_abspath_24603()

attributes #0 = { cold noreturn nounwind }
attributes #1 = { nocallback nofree nosync nounwind readnone speculatable willreturn }
```

into

```
pass:                                             ; preds = %top
  call void (ptr addrspace(10), ...) @julia.write_barrier_binding(ptr addrspace(10) null, ptr addrspace(10) %1)
  store i1 true, ptr undef, align 1
  ret void
```

The reason for that is this instcombine code that makes cc mismatches unreachable behavior:

https://github.com/llvm/llvm-project/blob/0c090dcc8a97a07bb3b3d2f64dbd1abf3990c1c6/llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp#L2897-L2924

That code is not new in LLVM15, so I don't know what triggered this to start happening now.

**Comment by Keno** [comment]
Side note for my own future reference. This was the reduce script for this:
```
#!/home/keno/julia-1.7.1/bin/julia
in = read(`/home/keno/julia-llvm15/usr/tools/llvm-extract --func=julia___init_build_28469 -S $(ARGS[1])`, String)
contains("ptr undef")(in) && exit(1)
contains("ptr null")(in) && exit(1)
contains("ptr addrspace(10) undef")(in) && exit(1)
out = try; read(pipeline(`/home/keno/julia-llvm15/usr/tools/opt -enable-new-pm=0 -load libjulia.so -juliaO3-sysimg $(ARGS[1]) -o -`, `/home/keno/julia-llvm15/usr/tools/llvm-extract --func=julia___init_build_28469 -S -`), String); catch; exit(1); end
#println(out)
contains("call void @jl_gc_queue_binding(ptr addrspace(10)")(out) || exit(1)
nu = length(findall("unreachable", out))
nu >= 2 && exit(0)
exit(1)
```

**Comment by Keno** [comment]
> That code is not new in LLVM15, so I don't know what triggered this to start happening now.

```
   %7 = call cc37 nonnull {}* bitcast ({}* ({}*, {}**, i32)* @j1_print_to_string_224 to {}* ({}*, {}*, {}*, {}*, {}*)*)({}* inttoptr (i64 140275598044992 to {}*), {}* inttoptr (i64 140275646630416 to {}*), {}* inttoptr (i64 140275594834608 to {}*), {}* inttoptr (i64 140275646630384 to {}*), {}* inttoptr (i64 140275623433424 to {}*))
```

Ah, see above, before opaque pointers, we had an extra bitcast there (so not a Function* directly), so the check here did not fire:
https://github.com/llvm/llvm-project/blob/0c090dcc8a97a07bb3b3d2f64dbd1abf3990c1c6/llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp#L2881-L2887

**Comment by Keno** [comment]
I think our choices here are:
1. Mark the callee cc37 also and change calling conventions at the same time in the final lowering
2. Teach LLVM that the two callings conventions are compatible. Generally the callee will have C calling convention, so we could re-use the existing hook in TLI, but I'm not sure it's overridable in the frontend.

Any other ideas?

**Comment by vtjnash** [comment]
IIRC, we mark it "thunk", since that declares that it is not permitted to assume the calling convention will match (yet)

**Comment by preames** [comment]
I'm missing some context to understand the lowering discussion, but the mismatched calling conventions is definitely invalid IR.  

I can't comment on Keno's first suggestion above as I don't understand the lowering mentioned.  I will note that assuming the lowering bits are manageable, changing the IR to be valid would be generally preferable.  

The second option would require an LLVM source change, but a) it'd be pretty straight forward, and b) you must already have a source change to define the calling convention.

**Comment by Keno** [comment]
So for us cc37 is semantically a calling convention that is compatible with the c signature `(void *f, void **args, size_t nargs)` which is how our runtime internally treats calls with all boxed arguments. In cc37, the arguments are passed individually rather than allocating the argument buffer on the stack. The reason not to emit the c signature directly is that LLVM is significantly better at optimizing the cc37 variant and also some of our GC lowering relies on not putting tracked pointers into allocas (though the latter could be fixed). The GC root placement pass rewrites the CC at the very end of IR lowering (and refuses the GC frame for the argument buffer).

Do you think it would be reasonable to
a) Reserve an upstream cc for our semantics and mark it as c compatible - swift had one, so I don't see why we couldn't
Or
b) Have LLVM treat all unknown ccs as C compatible?

We do need an upstream compatible solution, since we can't carry semantics-changing patches, since some downstream distros like to build us against upstream LLVM.

**Comment by Keno** [comment]
I had an extended discussion with @preames on what to do here and we discussed several options. I won't recount the entire discussion here, but we discussed a few possible options that I wanted to write down.

Option A) Implement a jlcall calling convention upstream, mark it as C-compatible and proceed as we have been.
Option B) Change all jlcall function signatures to be vararg with special julia vararg intrinsics, e.g.
```
define cc37 @foo_jlcall(...)
{
      %f = ptr call @julia.va.f()
      %args = ptr call @julia.va.args()
      %nargs = ptr call @julia.va.nargs()
}
```
Option C) Have an intrinsic that trades a function pointer for a cc37 function pointer. It's a bit ugly but @preames points out that there is nothing technically wrong with having to function pointers have the same pointer value, but be different function calls (I pointed to the PowerPC dual-entry cc as an example).
```
define ccc @bar()
{
%x = box(1)
%y = box(2)
%foo_37 = call ptr @julia.wrapper37(ptr @foo_jlcall)
call cc37 foo_37(ptr @foo, %x, %y)
}
```
Option D) Have an an indirect call intrinsic, along the lines of
```
call cc37 ptr @julia.call(ptr @foo_jlcall, ptr @foo, ptr %x, ptr %y)
```
Doesn't have to be cc37 at that point, because we can detect which things we need to lower by looking at the callee.

My plan is to try out Option D) and see if there's any issues with it. If so, then Option A) is something we could pursue, but the path to upstream is not entirely clear. Will post here as I have more data.
